File: frontend/src/app/services/live-coaching.service.ts
import { inject, Injectable, NgZone } from '@angular/core';
import { BehaviorSubject, Subscription } from 'rxjs';
import { distinctUntilChanged, filter, map } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { BluetoothService, LiveMetrics } from './bluetooth.service';
import { WorkoutExecutionService } from './workout-execution.service';
import { TrainingService } from './training.service';
import { parseSseBuffer } from './sse-parser.util';

export interface CoachingCue {
  message: string;
  timestamp: Date;
  triggerType: 'PERIODIC' | 'BLOCK_START' | 'BLOCK_END';
}

interface CoachingRequest {
  sportType: string;
  ftp: number;
  blockType: string;
  blockLabel: string;
  blockIndex: number;
  totalBlocks: number;
  targetIntensityPercent: number;
  remainingSeconds: number;
  blockDurationSeconds: number;
  avgPower: number;
  avgCadence: number;
  avgHeartRate: number;
  blockAvgPower: number;
  blockAvgCadence: number;
  blockAvgHR: number;
  sessionAvgPower: number;
  sessionAvgHR: number;
  triggerType: string;
}

const PERIODIC_INTERVAL_MS = 15_000;
const METRICS_WINDOW = 15;

@Injectable({ providedIn: 'root' })
export class LiveCoachingService {
  private bluetoothService = inject(BluetoothService);
  private executionService = inject(WorkoutExecutionService);
  private trainingService = inject(TrainingService);
  private ngZone = inject(NgZone);

  private cueSubject = new BehaviorSubject<CoachingCue | null>(null);
  cue$ = this.cueSubject.asObservable();

  private enabledSubject = new BehaviorSubject<boolean>(
    localStorage.getItem('ai-coaching-enabled') !== 'false',
  );
  enabled$ = this.enabledSubject.asObservable();

  private recentMetrics: LiveMetrics[] = [];
  private metricsSub: Subscription | null = null;
  private blockSub: Subscription | null = null;
  private periodicTimer: ReturnType<typeof setInterval> | null = null;
  private inFlight = false;
  private running = false;

  get enabled(): boolean {
    return this.enabledSubject.value;
  }

  toggle(): void {
    const next = !this.enabledSubject.value;
    this.enabledSubject.next(next);
    localStorage.setItem('ai-coaching-enabled', String(next));
    if (next && !this.running) {
      this.start();
    } else if (!next) {
      this.stop();
    }
  }

  start(): void {
    if (this.running) return;
    if (!this.enabledSubject.value) return;
    this.running = true;
    this.recentMetrics = [];

    // Collect metrics in a rolling window
    this.metricsSub = this.bluetoothService.metrics$.subscribe((m) => {
      this.recentMetrics.push(m);
      if (this.recentMetrics.length > METRICS_WINDOW) {
        this.recentMetrics.shift();
      }
    });

    // Detect block transitions
    this.blockSub = this.executionService.state$
      .pipe(
        map((s) => s.currentBlockIndex),
        distinctUntilChanged(),
      )
      .subscribe((blockIndex) => {
        if (blockIndex > 0) {
          this.requestCue('BLOCK_END');
        }
        if (blockIndex >= 0) {
          // Small delay so BLOCK_END fires first
          setTimeout(() => this.requestCue('BLOCK_START'), 500);
        }
      });

    // Periodic coaching every 15 seconds
    this.periodicTimer = setInterval(() => {
      const state = this.executionService.currentState;
      if (state.isActive && !state.isPaused && !this.inFlight) {
        this.requestCue('PERIODIC');
      }
    }, PERIODIC_INTERVAL_MS);
  }

  stop(): void {
    this.running = false;
    this.metricsSub?.unsubscribe();
    this.metricsSub = null;
    this.blockSub?.unsubscribe();
    this.blockSub = null;
    if (this.periodicTimer) {
      clearInterval(this.periodicTimer);
      this.periodicTimer = null;
    }
    this.recentMetrics = [];
    this.cueSubject.next(null);
  }

  private async requestCue(triggerType: string): Promise<void> {
    if (!this.enabledSubject.value || !this.running) return;

    const state = this.executionService.currentState;
    if (!state.training || !state.isActive) return;

    const block = state.flatBlocks[state.currentBlockIndex];
    if (!block) return;

    this.inFlight = true;

    const avg = this.aggregateMetrics();
    const ftp = this.trainingService.currentFtp ?? 200;

    const request: CoachingRequest = {
      sportType: state.training.sportType || 'CYCLING',
      ftp,
      blockType: block.type,
      blockLabel: block.label || block.type,
      blockIndex: state.currentBlockIndex,
      totalBlocks: state.flatBlocks.length,
      targetIntensityPercent: block.intensityTarget || 0,
      remainingSeconds: state.remainingBlockSeconds,
      blockDurationSeconds: block.durationSeconds || 0,
      avgPower: avg.power,
      avgCadence: avg.cadence,
      avgHeartRate: avg.heartRate,
      blockAvgPower: state.currentBlockAverages.power,
      blockAvgCadence: state.currentBlockAverages.cadence,
      blockAvgHR: state.currentBlockAverages.heartRate,
      sessionAvgPower: state.averages.power,
      sessionAvgHR: state.averages.heartRate,
      triggerType,
    };

    try {
      const message = await this.fetchCoachingCue(request);
      if (message && message.trim()) {
        this.ngZone.run(() => {
          this.cueSubject.next({
            message: message.trim(),
            timestamp: new Date(),
            triggerType: triggerType as CoachingCue['triggerType'],
          });
        });
      }
    } catch {
      // Silently ignore coaching errors — non-critical feature
    } finally {
      this.inFlight = false;
    }
  }

  private async fetchCoachingCue(request: CoachingRequest): Promise<string> {
    const jwt = localStorage.getItem('token');
    const headers: Record<string, string> = { 'Content-Type': 'application/json' };
    if (jwt) headers['Authorization'] = `Bearer ${jwt}`;

    const response = await fetch(`${environment.apiUrl}/api/ai/coaching/cue`, {
      method: 'POST',
      headers,
      body: JSON.stringify(request),
    });

    if (!response.ok) return '';

    const reader = response.body!.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let result = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const { events, remaining } = parseSseBuffer(buffer);
      buffer = remaining;

      for (const { eventType, data } of events) {
        if (eventType === 'coaching') {
          result += data;
        }
      }
    }

    return result;
  }

  private aggregateMetrics(): { power: number; cadence: number; heartRate: number } {
    if (this.recentMetrics.length === 0) {
      return { power: 0, cadence: 0, heartRate: 0 };
    }
    const len = this.recentMetrics.length;
    let power = 0;
    let cadence = 0;
    let heartRate = 0;
    for (const m of this.recentMetrics) {
      power += m.power;
      cadence += m.cadence;
      heartRate += m.heartRate || 0;
    }
    return {
      power: Math.round(power / len),
      cadence: Math.round(cadence / len),
      heartRate: Math.round(heartRate / len),
    };
  }
}

File: backend/src/main/java/com/koval/trainingplannerbackend/ai/LiveCoachingService.java
package com.koval.trainingplannerbackend.ai;

import org.springframework.ai.chat.client.ChatClient;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.codec.ServerSentEvent;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;

/**
 * Stateless, lightweight AI coaching during live workout sessions.
 * Each request is independent — no conversation memory, no tools.
 */
@Service
public class LiveCoachingService {

    private final ChatClient coachingClient;

    public LiveCoachingService(@Qualifier("liveCoachingClient") ChatClient coachingClient) {
        this.coachingClient = coachingClient;
    }

    public Flux<ServerSentEvent<String>> coach(CoachingRequest request) {
        String userMessage = buildCompactPrompt(request);
        return coachingClient.prompt()
                .user(userMessage)
                .stream()
                .content()
                .filter(token -> token != null && !token.isEmpty())
                .map(token -> ServerSentEvent.<String>builder()
                        .event("coaching")
                        .data(token)
                        .build())
                .concatWith(Flux.just(ServerSentEvent.<String>builder()
                        .event("done")
                        .data("")
                        .build()))
                .onErrorResume(ex -> Flux.just(ServerSentEvent.<String>builder()
                        .event("error")
                        .data("Coaching unavailable")
                        .build()));
    }

    private String buildCompactPrompt(CoachingRequest r) {
        var sb = new StringBuilder();
        sb.append("sport=").append(r.sportType()).append(" ftp=").append(r.ftp()).append('\n');
        sb.append("block=").append(r.blockType()).append(' ').append(r.blockIndex() + 1)
                .append('/').append(r.totalBlocks());
        if (r.blockLabel() != null) sb.append(" \"").append(r.blockLabel()).append('"');
        sb.append(" target=").append(r.targetIntensityPercent()).append('%');
        sb.append(" remaining=").append(r.remainingSeconds()).append("s");
        sb.append(" duration=").append(r.blockDurationSeconds()).append("s\n");
        sb.append("last15s: power=").append(r.avgPower())
                .append(" cad=").append(r.avgCadence())
                .append(" hr=").append(r.avgHeartRate()).append('\n');
        sb.append("block_avg: power=").append(r.blockAvgPower())
                .append(" cad=").append(r.blockAvgCadence())
                .append(" hr=").append(r.blockAvgHR()).append('\n');
        sb.append("session_avg: power=").append(r.sessionAvgPower())
                .append(" hr=").append(r.sessionAvgHR()).append('\n');
        sb.append("trigger=").append(r.triggerType());
        return sb.toString();
    }

    public record CoachingRequest(
            String sportType,
            int ftp,
            String blockType,
            String blockLabel,
            int blockIndex,
            int totalBlocks,
            int targetIntensityPercent,
            int remainingSeconds,
            int blockDurationSeconds,
            int avgPower,
            int avgCadence,
            int avgHeartRate,
            int blockAvgPower,
            int blockAvgCadence,
            int blockAvgHR,
            int sessionAvgPower,
            int sessionAvgHR,
            String triggerType
    ) {}
}
